mongodb-plugin/src/main/java/io/cdap/plugin/batch/source/MongoDBBatchSource.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.batch.source;
+
+import com.google.common.base.Strings;
+import com.mongodb.hadoop.MongoInputFormat;
+import com.mongodb.hadoop.splitter.MongoSplitter;
+import com.mongodb.hadoop.splitter.StandaloneMongoSplitter;
+import com.mongodb.hadoop.util.MongoConfigUtil;
+import io.cdap.cdap.api.annotation.Description;
+import io.cdap.cdap.api.annotation.Macro;
+import io.cdap.cdap.api.annotation.Name;
+import io.cdap.cdap.api.annotation.Plugin;
+import io.cdap.cdap.api.data.batch.Input;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.api.dataset.lib.KeyValue;
+import io.cdap.cdap.etl.api.Emitter;
+import io.cdap.cdap.etl.api.PipelineConfigurer;
+import io.cdap.cdap.etl.api.batch.BatchRuntimeContext;
+import io.cdap.cdap.etl.api.batch.BatchSource;
+import io.cdap.cdap.etl.api.batch.BatchSourceContext;
+import io.cdap.plugin.BSONConverter;
+import io.cdap.plugin.MongoDBConfig;
+import io.cdap.plugin.MongoDBConstants;
+import io.cdap.plugin.common.LineageRecorder;
+import io.cdap.plugin.common.ReferenceBatchSource;
+import io.cdap.plugin.common.ReferencePluginConfig;
+import io.cdap.plugin.common.SourceInputFormatProvider;
+import org.apache.hadoop.conf.Configuration;
+import org.bson.BSONObject;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link BatchSource} that reads data from MongoDB and converts each document into 
+ * a {@link StructuredRecord} with the help of the specified Schema.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name(MongoDBConstants.PLUGIN_NAME)
+@Description("MongoDB Batch Source will read documents from MongoDB and convert each document " +
+  "into a StructuredRecord with the help of the specified Schema. ")
+public class MongoDBBatchSource extends ReferenceBatchSource<Object, BSONObject, StructuredRecord> {
+
+  private final MongoDBSourceConfig config;
+  private BSONConverter bsonConverter;
+
+  public MongoDBBatchSource(MongoDBSourceConfig config) {
+    super(new ReferencePluginConfig(config.referenceName));
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    Schema schema = config.getSchema();
```

Comment:
Schema may be null on this stage

Reply:
[Changed](https://github.com/vladhlinskiy/database-plugins/pull/14/files#diff-56e5c8c248fc520b3c862df33668d2feR76) to validate the config first:

```
  @Override
  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
    super.configurePipeline(pipelineConfigurer);
    config.validate();
    Schema schema = config.getSchema();
    ...
```